### PR TITLE
Zero out unused bytes in snapshots

### PR DESCRIPTION
**Pull request by martijnthe**
Compiled code blocks are sized in multiples of JMEM_ALIGNMENT,
but it's possible that some bytes at the end remain unused and get filled
with junk. This causes snapshot output to become nondeterministic.
To fix this, zero out the compiled code buffer before using it.

JerryScript-DCO-1.0-Signed-off-by: Martijn The martijn.the@intel.com

### CODE REVIEW

**Comment by martijnthe** [comment]
cc @zherczeg @jiangzidong

**Comment by jiangzidong** [comment]
LGTM (informal)

**Comment by zherczeg** [comment]
I would like to land #1797 first.

File: jerry-core/parser/js/js-parser.c [hunk 1]
```diff
@@ -1600,6 +1600,11 @@ parser_post_processing (parser_context_t *context_p) /**< context */
 
   compiled_code_p = (ecma_compiled_code_t *) parser_malloc (context_p, total_size);
 
+#ifdef JERRY_ENABLE_SNAPSHOT_SAVE
+  // Avoid getting junk bytes at the end when bytes at the end remain unused:
+  memset (compiled_code_p, 0, total_size);
```
Comment (zherczeg) [inline jerry-core/parser/js/js-parser.c]: If only the last few bytes are unused, why don't we zero the last block only?

File: tests/unit-core/test-api.c [hunk 2]
```diff
@@ -1169,6 +1169,20 @@ main (void)
...
+    /* Check the snapshot data. Unused bytes should be filled with zeroes */
+    const uint8_t expected_data[] =
+    {
+      0x07, 0x00, 0x00, 0x00, 0x38, 0x00, 0x00, 0x00, 0x20, 0x00, 0x00, 0x00, 0x01,
+      0x00, 0x00, 0x00, 0x03, 0x00, 0x01, 0x00, 0x01, 0x00, 0x01, 0x00, 0x00, 0x00,
+      0x00, 0x01, 0x05, 0x00, 0x28, 0x00, 0xB7, 0x46, 0x00, 0x00, 0x00, 0x00, 0x00,
+      0x00, 0x02, 0x00, 0x01, 0x00, 0x21, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01, 0x01,
+      0x01, 0x00, 0x47, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x14,
+      0x00, 0x73, 0x74, 0x72, 0x69, 0x6E, 0x67, 0x20, 0x66, 0x72, 0x6F, 0x6D, 0x20,
+      0x73, 0x6E, 0x61, 0x70, 0x73, 0x68, 0x6F, 0x74, 0x00, 0x00
```
Comment (zherczeg) [inline tests/unit-core/test-api.c]: What about big endian systems?

Comment (martijnthe) [reply tests/unit-core/test-api.c]: Hmm, I don't know... Any suggestions?
Could use `#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__` etc.
If possible I'd rather only test for the zero padding, but I don't see a way to do that w/o using the internal structures here directly.

File: jerry-core/parser/js/js-parser.c [hunk 3]
```diff
@@ -1596,10 +1599,21 @@ parser_post_processing (parser_context_t *context_p) /**< context */
...
 
   compiled_code_p = (ecma_compiled_code_t *) parser_malloc (context_p, total_size);
 
+#ifdef JERRY_ENABLE_SNAPSHOT_SAVE
+  // Avoid getting junk bytes at the end when bytes at the end remain unused:
+  if (total_size_used < total_size)
+  {
+    memset (((uint8_t *) compiled_code_p) + total_size_used, 0, total_size - total_size_used);
+  }
+#endif
```
Comment (martijnthe) [inline jerry-core/parser/js/js-parser.c]: @zherczeg better like this?

**Comment by akosthekiss** [comment]
#1797 is in master now, I think this PR has green light to progress now (but needs rebasing and conflict resolution)

**Comment by LaszloLango** [comment]
@martijnthe please update or close this PR.

**Comment by LaszloLango** [comment]
@martijnthe, could you update the PR?

**Comment by martijnthe** [comment]
@LaszloLango rebased & updated it. Sorry I missed your earlier message.

**Comment by LaszloLango** [comment]
@martijnthe please fix the style issue, so we can land your PR.

**Comment by martijnthe** [comment]
Fixed (hopefully)

**Pull request merged**